python/lsst/cp/pipe/measureCrosstalk.py

```diff
@@ -0,0 +1,799 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import itertools
+import numpy as np
+from scipy.stats import norm
+
+from collections import defaultdict
+
+import lsst.pipe.base as pipeBase
+import lsst.pipe.base.connectionTypes as cT
+
+from lsstDebug import getDebugFrame
+from lsst.afw.detection import FootprintSet, Threshold
+from lsst.afw.display import getDisplay
+from lsst.pex.config import Config, Field, ListField, ConfigurableField
+from lsst.ip.isr import (CrosstalkCalib, IsrProvenance)
+from lsst.pipe.tasks.getRepositoryData import DataRefListRunner
+
+__all__ = ["CrosstalkExtractConfig", "CrosstalkExtractTask",
+           "CrosstalkSolveTask", "CrosstalkSolveConfig",
+           "MeasureCrosstalkConfig", "MeasureCrosstalkTask"]
+
+
+class CrosstalkExtractConnections(pipeBase.PipelineTaskConnections,
+                                  dimensions=("instrument", "exposure", "detector")):
+    inputExp = cT.Input(
+        name="crosstalkInputs",
+        doc="Input post-ISR processed exposure to measure.",
+        storageClass="Exposure",
+        dimensions=("instrument", "exposure", "detector"),
+        multiple=False,
+    )
+    # TODO: Depends on DM-21904.
+    sourceExp = cT.Input(
+        name="crosstalkSource",
+        doc="Input post-ISR processed exposure to measure for impact on inputExp.",
+        storageClass="Exposure",
+        dimensions=("instrument", "exposure", "detector"),
+        multiple=True,
+        deferLoad=True,
+        # lookupFunction=None,
+    )
+
+    outputRatios = cT.Output(
+        name="crosstalkRatios",
+        doc="Extracted crosstalk pixel ratios.",
+        storageClass="StructuredDataDict",
+        dimensions=("instrument", "exposure", "detector"),
+    )
+    outputFluxes = cT.Output(
+        name="crosstalkFluxes",
+        doc="Source pixel fluxes used in ratios.",
+        storageClass="StructuredDataDict",
+        dimensions=("instrument", "exposure", "detector"),
+    )
+
+    def __init__(self, *, config=None):
+        super().__init__(config=config)
+        self.inputs.discard("sourceExp")
+
+
+class CrosstalkExtractConfig(pipeBase.PipelineTaskConfig,
+                             pipelineConnections=CrosstalkExtractConnections):
+    """Configuration for the measurement of pixel ratios.
+    """
+    doMeasureInterchip = Field(
+        dtype=bool,
+        default=False,
+        doc="Measure inter-chip crosstalk as well?",
+    )
+    threshold = Field(
+        dtype=float,
+        default=30000,
+        doc="Minimum level of source pixels for which to measure crosstalk."
+    )
+    badMask = ListField(
+        dtype=str,
+        default=["SAT", "BAD", "INTRP"],
```

Comment:
If the crosstalk happens after the ADC, then they probably should be retained.  If it happens before, then some fraction of a potentially unknown source level is contributing to the target signal.

Reply:
I'm not sure I agree with that exact phrasing, but I certainly agree with the spirit that it needs to be configurable. I think it's more that if the ADC can digitise the full-well value then you should keep and otherwise you must mask.

However, the LSST system will definitely be able to digitise the full-well (both spot and flat full wells for that matter), so I think the default should be without `SAT`.

Reply:
To go one step further, it would actually be really nice to have a separate config option spelling this out for users, as it's not really conceptually in the same basket as the others IMO. Something like `ignoreSaturatedPixels=False` in the config, and then in the body just `if ignoreSaturatedPixels is True and 'SAT' not in badMask: badMask.append('SAT)`. What do you think of that approach?

Reply:
I'm not too happy with having the same thing configured in two different places, but I've added the option, and then added a `config.validate` method that should append or filter the `badMask` field to match.  This at least forces the config to be consistent, so there shouldn't be too much additional confusion.

Reply:
OK, that sounds good to me.